third_party/move/move-vm/runtime/src/move_vm.rs

```diff
@@ -101,17 +126,12 @@ impl MoveVM {
     /// outstanding sessions created from this VM.
     /// TODO: new loader architecture
     pub fn flush_loader_cache_if_invalidated(&self) {
+        if self.runtime.loader().is_invalidated() {
```

Comment:
if invalidated, cache is flushed, but the next line says `self.runtime.loader().flush_if_invalidated()`, which seems confusing?

Reply:
Yea this is indeed a bit confusing. The back story of this is following: in order to keep the public api compatiblity, I still put a code cache inside the MoveVM and this code cache will be used as the default implementor of the trait. However, this makes module upgrade to be a bit tricky: if there's a module upgrade, our old logic will require us to flush all modules in the module cache. With the new logic however, the caller can decide whether to keep the cache or not during upgrade. My point is that the logic here is tightly coupled with how we will be treating module upgrade in the future and we would probably get a bit more clarity on this once we try to version the module cache with MVHashMap. Thus the logic here is just for correctness for now and we might need to adjust the interfaces there in the future.

Reply:
Could it be possibly cleaner to have an explicit enum in module_cache (also helpful to e.g. have a local cache in front in the new way when Arc points to a remote concurrent cache), and flush when V0 (old way) and not in the V1 (remote -> MVHashMap)

Reply:
Let's also add a comment in any case.